### PR TITLE
8326957: Implement JEP 474: ZGC: Generational Mode by Default

### PR DESCRIPTION
**Pull request by xmas92**
This is the implementation task for `JEP 474: ZGC: Generational Mode by Default`. See the JEP for details. [JDK-8326667](https://bugs.openjdk.org/browse/JDK-8326667)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8326958](https://bugs.openjdk.org/browse/JDK-8326958) to be approved

### Issues
 * [JDK-8326957](https://bugs.openjdk.org/browse/JDK-8326957): Implement JEP 474: ZGC: Generational Mode by Default (**Sub-task** - P4)
 * [JDK-8326958](https://bugs.openjdk.org/browse/JDK-8326958): Implement JEP 474: ZGC: Generational Mode by Default (**CSR**)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [ca9492d9](https://git.openjdk.org/jdk/pull/18393/files/ca9492d9951ed1175494ad60f99535faf1fb506e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18393/head:pull/18393` \
`$ git checkout pull/18393`

Update a local copy of the PR: \
`$ git checkout pull/18393` \
`$ git pull https://git.openjdk.org/jdk.git pull/18393/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18393`

View PR using the GUI difftool: \
`$ git pr show -t 18393`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18393.diff">https://git.openjdk.org/jdk/pull/18393.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18393#issuecomment-2046812295)

### CODE REVIEW

File: src/hotspot/share/gc/x/xInitialize.cpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ XInitialize::XInitialize(XBarrierSet* barrier_set) {
   log_info(gc, init)("Version: %s (%s)",
                      VM_Version::vm_release(),
                      VM_Version::jdk_debug_level());
-  log_info(gc, init)("Using legacy single-generation mode");
+  log_info(gc, init)("Using deprecated single-generation mode");
```
Comment (fisk) [inline src/hotspot/share/gc/x/xInitialize.cpp:44]: Change to non-generational instead of single generation, for consistency.

**Review by fisk — APPROVED** [review]
Looks good. One nit in terminology, don't need to see the new version.

**Review by stefank — APPROVED** [review]

**Comment by tkrodriguez** [comment]
Graal still doesn't support generational ZGC though I'm actively working on it.  I'm hoping to have it in before rampdown but at least until that time JVMCI needs to default to non-generational ZGC.

**Comment by stefank** [comment]
> Graal still doesn't support generational ZGC though I'm actively working on it. I'm hoping to have it in before rampdown but at least until that time JVMCI needs to default to non-generational ZGC.

Great that you are working on support for Generational ZGC!

The intent of this JEP is to make it so that you get Generational ZGC when you specify -XX:+UseZGC, and then warn that non-generational ZGC is deprecated and warn if the user explicitly turns it on. I don't want us to change the meaning of -XX:+UseZGC when the Graal JIT is enabled. Doing so would be inconsistent, cause confusion, and could be misleading to our users.

In fact, we have already seen instances of a somewhat opposite problem. Today if you try to run Generational ZGC with the Graal JIT, you get Generational ZGC but revert back to using C2. We print a warning when that happens, but we have seen users not realizing that they ran C2 instead of the Graal JIT.

My proposal is to simply refuse to start the JVM and print an error message when we have conflicting GC and JIT compiler flags.

With this we would have:
* -XX:+UseZGC -XX:<Use Graal JIT flags> => Refuse to start with error message
* -XX:+UseZGC -XX:-ZGenerational -XX:<Use Graal JIT flags> => Same as above
* -XX:+UseZGC -XX:-ZGenerational -XX:<Use Graal JIT flags> => Start (with message about deprecation of non-generational ZGC)

With this we would both get rid of the confusing situation where the user asks for Graal JIT but get C2 and the other confusing situation where the user asks for Generational ZGC but gets non-generational ZGC. The user will have to make a decision on what combination they want to run.

**Comment by stefank** [comment]
This is my proposal:

For this JEP, I propose that we stick with the current plan that -XX:+UseZGC always mean Generational ZGC. If the user try to enable the Graal JIT when running Generational ZGC, we'll print a warning and switch over to C2, just like we're doing today.

Then, as a follow-up, we can figure out what the correct behavior should be if the user specifies an incompatible GC and compiler combination.

@tkrodriguez Is this an OK action plan for you?

**Comment by tkrodriguez** [comment]
Sorry I didn't reply sooner.  That sounds like a fine interim solution.  I agree that automatically selecting something different can be very confusing.  Anyway, hopefully this kind of problem will be a transient one.

**Comment by stefank** [comment]
Thanks!

**Comment by dougxc** [comment]
After discussing with Tom offline, here's one more change you could make in this PR:
* Change the warning in `check_jvmci_supported_gc` to a hard error when an unsupported GC + compiler combination is requested.

Either way, as soon we have Gen ZGC implemented for Graal, we will:
1. Update `gc_supports_jvmci`.
2. Update `check_jvmci_supported_gc` to no longer warn and switch compilers but to exit the VM when an unsupported combination is requested.

**Review by stefank — APPROVED** [review]

**Comment by xmas92** [comment]
Thanks for the reviews.
/integrate

**Pull request closed without merging**